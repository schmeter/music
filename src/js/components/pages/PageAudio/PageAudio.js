import React from 'react';
import PropTypes from 'prop-types';

import Link from '../../Link';
import Page from '../../Page';
import { Page404 } from '../';
import AlbumList from './AlbumList';
import { getUrl } from '../../../services/navigation';
import { joinTitleParts, getTitle } from '../../../services/meta';

class PageAudio extends React.Component {
    getTitle() {
        const titleParts = [];
        const { match, getArtist, getAlbum } = this.props;

        const { artistId, albumId } = match.params;
        const selectedArtist = getArtist(artistId);

        const selectedAlbum = getAlbum(artistId, albumId);

        if (selectedArtist) {
            titleParts.push(selectedArtist.title);
        }
        if (selectedAlbum) {
            titleParts.push(selectedAlbum.title);
        }

        return getTitle(joinTitleParts(titleParts));
    }

    render() {
        const { albums, match, getArtist, getAlbum } = this.props;

        const { artistId, albumId } = match.params;
        const selectedArtist = getArtist(artistId);
        const selectedAlbum = getAlbum(artistId, albumId);
        const validParams = (artistId ? selectedArtist : true)
            && (albumId ? selectedAlbum : true);

        return !validParams ? <Page404 /> : (
            <Page
                id="audio"
                title={this.getTitle()}
                useBaseClass={false}
            >
                <AlbumList
                    albumList={albums}
                    selectedArtist={selectedArtist}
                    selectedAlbum={selectedAlbum}
                />
                <div className="center secret">
                    <Link to={getUrl('settings')}>.</Link>
                </div>
            </Page>
        );
    }
}

PageAudio.propTypes = {
    albums: PropTypes.array.isRequired,
    match: PropTypes.object.isRequired,
    getArtist: PropTypes.func.isRequired,
    getAlbum: PropTypes.func.isRequired
};

export default PageAudio;
