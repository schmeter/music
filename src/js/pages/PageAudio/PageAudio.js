import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';

import { Page404 } from '../';
import Page from '../../components/Page';
import AlbumList from './AlbumList';
import { joinTitleParts, setTitle } from '../../services/meta';

const PageAudio = ({
    getArtist,
    getAlbum,
}) => {
    const { artistId, albumId } = useParams();
    const selectedArtist = getArtist(artistId);
    const selectedAlbum = getAlbum(artistId, albumId);
    const validParams = (!artistId || selectedArtist) && (!albumId || selectedAlbum);

    useEffect(() => {
        const titleParts = [];

        if (selectedArtist) {
            titleParts.push(selectedArtist.title);
        }
        if (selectedAlbum) {
            titleParts.push(selectedAlbum.title);
        }

        setTitle(joinTitleParts(titleParts));
    }, [
        selectedArtist,
        selectedAlbum,
    ]);

    return !validParams ? <Page404 /> : (
        <Page
            id="audio"
            useBaseClass={false}
        >
            <AlbumList
                selectedArtist={selectedArtist}
                selectedAlbum={selectedAlbum}
            />
        </Page>
    );
};

PageAudio.propTypes = {
    getArtist: PropTypes.func.isRequired,
    getAlbum: PropTypes.func.isRequired,
};

export default PageAudio;
