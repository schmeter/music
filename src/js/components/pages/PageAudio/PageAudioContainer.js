import { connect } from 'react-redux';

import PageAudio from './PageAudio';
import { getArtist, getAlbum, getAlbums } from '../../../state/selectors/audio';

const mapStateToProps = (state) => ({
    albums: getAlbums(state),
    getArtist: (artistId) => getArtist(state, artistId),
    getAlbum: (artistId, albumId) => getAlbum(state, artistId, albumId),
});

export default connect(mapStateToProps, null)(PageAudio);
