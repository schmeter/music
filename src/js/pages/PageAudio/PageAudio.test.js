import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import PageAudio from './PageAudio';

let mockUrlParams;

jest.mock('react-router-dom', () => ({
    useParams: () => (mockUrlParams),
}));
// eslint-disable-next-line react/display-name
jest.mock('../../components/Page', () => () => <div />);
// eslint-disable-next-line react/display-name
jest.mock('./AlbumList', () => () => <div />);
jest.mock('../', () => ({
    // eslint-disable-next-line react/display-name
    Page404: () => <div />,
}));

describe('PageAudio', () => {
    it('renders correctly as index page without url params', () => {
        mockUrlParams = {};
        const props = {
            getArtist: () => false,
            getAlbum: () => false,
        };
        const component = mount(<PageAudio {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('renders correctly with 1 url param', () => {
        mockUrlParams = {
            artistId: 'test',
        };
        const props = {
            getArtist: () => true,
            getAlbum: () => false,
        };
        const component = mount(<PageAudio {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('renders correctly with 2 url params', () => {
        mockUrlParams = {
            artistId: 'test',
            albumId: 'test',
        };
        const props = {
            getArtist: () => true,
            getAlbum: () => true,
        };
        const component = mount(<PageAudio {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('renders correctly with 1 wrong url param', () => {
        mockUrlParams = {
            artistId: 'test',
        };
        const props = {
            getArtist: () => false,
            getAlbum: () => false,
        };
        const component = mount(<PageAudio {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });

    it('renders correctly with 2 wrong url params', () => {
        mockUrlParams = {
            artistId: 'test',
            albumId: 'test',
        };
        const props = {
            getArtist: () => false,
            getAlbum: () => false,
        };
        const component = mount(<PageAudio {...props} />);

        expect(toJson(component)).toMatchSnapshot();
    });
});
