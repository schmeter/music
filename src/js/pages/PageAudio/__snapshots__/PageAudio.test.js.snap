// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageAudio renders correctly as index page without url params 1`] = `
<PageAudio
  getAlbum={[Function]}
  getArtist={[Function]}
  isIndexPage={true}
>
  <Component
    id="audio"
    useBaseClass={false}
  >
    <div />
  </Component>
</PageAudio>
`;

exports[`PageAudio renders correctly with 1 url param 1`] = `
<PageAudio
  getAlbum={[Function]}
  getArtist={[Function]}
>
  <Component
    id="audio"
    useBaseClass={false}
  >
    <div />
  </Component>
</PageAudio>
`;

exports[`PageAudio renders correctly with 1 wrong url param 1`] = `
<PageAudio
  getAlbum={[Function]}
  getArtist={[Function]}
>
  <Page404>
    <div />
  </Page404>
</PageAudio>
`;

exports[`PageAudio renders correctly with 2 url params 1`] = `
<PageAudio
  getAlbum={[Function]}
  getArtist={[Function]}
>
  <Component
    id="audio"
    useBaseClass={false}
  >
    <div />
  </Component>
</PageAudio>
`;

exports[`PageAudio renders correctly with 2 wrong url params 1`] = `
<PageAudio
  getAlbum={[Function]}
  getArtist={[Function]}
>
  <Page404>
    <div />
  </Page404>
</PageAudio>
`;
